File: finance-ui/src/types/schema.ts
/**
 * TypeScript type definitions for schemas
 */

export enum WorkType {
  DATA_PREPARATION = 'data_preparation',
  RECONCILIATION = 'reconciliation',
}

export enum SchemaStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
}

export interface Schema {
  id: number;
  user_id: number;
  name_cn: string;
  type_key: string;
  work_type: WorkType;
  schema_path: string;
  config_path: string;
  version: string;
  status: SchemaStatus;
  is_public: boolean;
  callback_url?: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface SchemaDetail extends Schema {
  schema_content?: Record<string, any>;
}

export interface CreateSchemaRequest {
  name_cn: string;
  work_type: WorkType;
  callback_url?: string;
  description?: string;
}

export interface UpdateSchemaRequest {
  name_cn?: string;
  schema_content?: Record<string, any>;
  status?: SchemaStatus;
  callback_url?: string;
  description?: string;
}

export interface SchemaListResponse {
  total: number;
  schemas: Schema[];
}

export interface SchemaState {
  schemas: Schema[];
  currentSchema: SchemaDetail | null;
  loading: boolean;
  fetchSchemas: (filters?: SchemaFilters) => Promise<void>;
  createSchema: (data: CreateSchemaRequest) => Promise<Schema>;
  updateSchema: (id: number, data: UpdateSchemaRequest) => Promise<Schema>;
  deleteSchema: (id: number) => Promise<void>;
  getSchema: (id: number) => Promise<SchemaDetail>;
}

export interface SchemaFilters {
  work_type?: WorkType;
  status?: SchemaStatus;
  skip?: number;
  limit?: number;
}

File: finance-mcp/api/models/schema.py
"""
UserSchema database model
"""
from sqlalchemy import Column, Integer, String, DateTime, Enum, Boolean, Text, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from api.database import Base
import enum


class WorkType(str, enum.Enum):
    DATA_PREPARATION = "data_preparation"
    RECONCILIATION = "reconciliation"


class SchemaStatus(str, enum.Enum):
    DRAFT = "draft"
    PUBLISHED = "published"


class UserSchema(Base):
    __tablename__ = "user_schemas"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name_cn = Column(String(100), nullable=False, comment="中文名称")
    type_key = Column(String(50), nullable=False, index=True, comment="类型键（英文标识）")
    work_type = Column(Enum(WorkType), nullable=False, comment="工作类型")
    schema_path = Column(String(500), nullable=False, comment="Schema JSON文件路径")
    config_path = Column(String(500), nullable=False, comment="配置文件路径")
    version = Column(String(20), default="1.0", comment="Schema版本")
    status = Column(Enum(SchemaStatus), default=SchemaStatus.DRAFT, comment="状态")
    is_public = Column(Boolean, default=False, comment="是否公开")
    callback_url = Column(String(500), nullable=True, comment="回调URL")
    description = Column(Text, nullable=True, comment="描述")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationship
    user = relationship("User", back_populates="schemas")

    def __repr__(self):
        return f"<UserSchema(id={self.id}, name_cn='{self.name_cn}', type_key='{self.type_key}', work_type='{self.work_type}')>"
